akka-actor-tests/src/test/scala/akka/util/BoundedBlockingQueueSpec.scala

```diff
@@ -0,0 +1,818 @@
+
+import scala.collection.mutable
+import scala.concurrent.{ Await, ExecutionContext, ExecutionContextExecutor, Future }
+import scala.util.control.Exception
+import DefaultExecutionContext._
+
+import collection.JavaConverters._
+
+class BoundedBlockingQueueSpec
+  extends WordSpec
+  with Matchers
+  with QueueSetupHelper
+  with CustomContainsMatcher
+  with BlockingHelpers {
+
+  import QueueTestEvents._
+
+  "Constructor" must {
+    "not accept a null backing Queue" in {
+      intercept[IllegalArgumentException] {
+        new BoundedBlockingQueue(10, null)
+      }
+    }
+
+    "not accept a non-empty backing Queue" in {
+      intercept[IllegalArgumentException] {
+        val listQ = new util.LinkedList[String]()
+        listQ.add("Hello")
+        new BoundedBlockingQueue[String](10, listQ)
+      }
+    }
+
+    "accept and empty backing Queue" in {
+      new BoundedBlockingQueue(10, new util.LinkedList()).size() should equal(0)
+    }
+
+    "get a positive maxCapacity" in {
+      intercept[IllegalArgumentException] {
+        new BoundedBlockingQueue(0, new util.LinkedList())
+      }
+    }
+
+    "check remainingCapacity for BlockingQueue" in {
+      intercept[IllegalArgumentException] {
+        new BoundedBlockingQueue(0, new LinkedBlockingQueue(1))
+      }
+      new BoundedBlockingQueue(1, new LinkedBlockingQueue(1)).remainingCapacity() should equal(1)
+    }
+
+  }
+
+  "remainingCapacity" must {
+    "check the backing queue size" in {
+      val TestContext(queue, events, _, _, _, backingQueue) = newBoundedBlockingQueue(1)
+
+      backingQueue.offer("1")
+      queue.remainingCapacity() shouldBe 0
+      events should contain(getSize)
+    }
+  }
+
+  "put" must {
+    "reject null elements" in {
+      val queue = newBoundedBlockingQueue(1).queue
+      intercept[NullPointerException] {
+        queue.put(null)
+      }
+    }
+
+    "call the backing queue if not full" in {
+      val TestContext(queue, events, _, _, _, _) = newBoundedBlockingQueue(2)
+      queue.put("Hello")
+      queue.put("World")
+      events should contain inOrder (offer("Hello"), offer("World"))
+    }
+
+    "signal notEmpty when an element is inserted" in {
+      val TestContext(queue, events, _, _, _, _) = newBoundedBlockingQueue(1)
+      queue.put("Hello")
+      events should contain(signalNotEmpty)
+    }
+
+    "block when the queue is full" in {
+      val TestContext(queue, events, _, _, _, _) = newBoundedBlockingQueue(1)
+      queue.offer("1")
+
+      mustBlockFor(100 milliseconds) {
+        queue.put("2")
+      }
+      events should contain inOrder (offer("1"), awaitNotFull)
+      events should not contain offer("2")
+    }
+
+    "block until the backing queue has space" in {
+      val TestContext(queue, events, _, _, _, _) = newBoundedBlockingQueue(1)
+      queue.offer("a")
+
+      val f = Future(queue.put("b"))
+
+      after(10 milliseconds) {
+        queue.take()
```

Comment:
Perhaps asserting `f.isCompleted should be(false)` before taking would make this test slightly more readable?